### PR TITLE
Hackathon subscription activation and registry service account implementation

### PR DESCRIPTION
**Pull request by dgolovin**

### CODE REVIEW

**Review by benoitf — COMMENTED** [review]
I have a question around all generated classes

if they're automatically generated, should we avoid to store them into the git registry and generate them as part of a npm task ?

**Comment by dgolovin** [comment]
@benoitf we can definitely do that, but they not going to change much and also would be good to have generated sources tagged as everything else.

Another option is to have npm package per API so other projects, like OpenShift Toolkit, can use it too.

**Comment by benoitf** [comment]
I think that your suggestion with another repository with only the generated classes would be the best option.

- it avoids to include generated stuff in this extension (as users/developers might be tempted to tweak some classes for their needs)
- it allows to reuse the model in different locations and lifecycle here is specific to the model change
- we never commit generated objects in the repository (in this 3rd party repository), only the model, tooling/scripts to generate it so again there is no risk to have manual edits

**Comment by vrothberg** [comment]
@benoitf @dgolovin can you give some pointers on how I can test this code?  I am blank on how the toolchain.

**Comment by vrothberg** [comment]
Thanks a lot, @benoitf !  The instruction were _very_ helpful.

**Comment by vrothberg** [comment]
Nothing to worry about right now but something to improve later on:  the service-account user names can be very long as seen on the screenshot below.  Maybe we can truncate them?

image

**Comment by benoitf** [comment]
@vrothberg if you can file an enhancement request at https://github.com/containers/podman-desktop/issues/new?assignees=&labels=kind%2Fenhancement+%E2%9C%A8&projects=&template=enhancement.yml

**Comment by vrothberg** [comment]
Thanks! Opened https://github.com/containers/podman-desktop/issues/5670****

**Comment by vrothberg** [comment]
One more thing I noticed while playing:

The Authentication page displayed "LOGGED OUT" after restarting Podman Desktop.  However, we're still logged into the registry and subscription-manager is still working inside the VM (even after reboots).  So we need to flesh out what "logged in" and what "logged out" means.

Screenshot 2024-01-24 at 14 40 33

**Comment by vrothberg** [comment]
Is this ready to merge?  Then we can cut an alpha release and incrementally improve.

File: package.json [hunk 1]
```diff
@@ -38,11 +41,12 @@
     "@types/js-yaml": "^4.0.5",
     "@vitest/coverage-v8": "^1.2.1",
     "mkdirp": "^2.1.3",
+    "openapi-typescript-codegen": "^0.27.0",
```
Comment (benoitf) [inline package.json:44]: I guess you don't need the code generator anymore

File: src/extension.ts [hunk 2]
```diff
@@ -18,121 +18,263 @@
...
   }
   return authService;
 }
 
+function parseJwt (token: string) {
+  var base64Url = token.split('.')[1];
+  var base64 = base64Url.replace(/-/g, '+').replace(/_/g, '/');
+  var jsonPayload = decodeURIComponent(Buffer.from(base64, 'base64').toString().split('').map(function(c) {
+      return '%' + ('00' + c.charCodeAt(0).toString(16)).slice(-2);
+  }).join(''));
```
Comment (benoitf) [inline src/extension.ts:56]:
```suggestion
  var jsonPayload = decodeURIComponent(Buffer.from(base64, 'base64').toString().split('').map((c) => {
      return '%' + ('00' + c.charCodeAt(0).toString(16)).slice(-2);
  }).join(''));
```

File: src/podman-cli.ts [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ***********************************************************************/
+import { isMac, isWindows } from './util';
```
Comment (benoitf) [inline src/podman-cli.ts:18]: if it needs to stay even after podman 5.0 I think you should not have duplicates of the podman extension code there

probably we would need to see how podman extension can export an API that this extension could consume

**Review by benoitf — COMMENTED** [review]
I think all github workflows, dependency updates, tsconfig changes could be separated small PR as they're not related to the changes

**Comment by dgolovin** [comment]
Closed in favor of 0.0.2-alpah.x branch PR https://github.com/redhat-developer/podman-desktop-redhat-account-ext/pull/34

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,22 @@
+FROM node:17.7-alpine3.14 AS extension-builder
```
[the remark this answers is not in the document]
Comment (dgolovin) [reply Dockerfile:1]: Removed in https://github.com/redhat-developer/podman-desktop-redhat-account-ext/pull/35

File: package.json (hunk 1 above)
Comment (dgolovin) [reply package.json:44]: resolved in https://github.com/redhat-developer/podman-desktop-redhat-account-ext/pull/35

File: src/extension.ts (hunk 2 above)
Comment (dgolovin) [reply src/extension.ts:56]: resolved in #35

File: src/podman-cli.ts (hunk 3 above)
Comment (dgolovin) [reply src/podman-cli.ts:18]: moved conversation to [#35](https://github.com/redhat-developer/podman-desktop-redhat-account-ext/pull/35/files#r1487218477)

Comment (dgolovin) [reply src/podman-cli.ts:18]: tsconfig.json fixed and contains only required changes https://github.com/redhat-developer/podman-desktop-redhat-account-ext/pull/35/files#r1487228251

**Pull request closed without merging**